### ISSUE TITLE
Unlike RemoteWebDriver, AppiumDriver constructors don't allow user-provided CommandExecutor

### ISSUE THREAD

**Issue by baechul** [appium/java-client#276]
There is a usecase where I need to have an AppiumCommandExecutor constructed with my HttpClient.Factory then pass it to the AppiumDriver constructors. Not limited to this case only but the usecase I have is to add some extra headers to the underlaying RemoteWebDriver HttpRequests by overriding HttpClient execute method in my HttpClient.Factory.

Some enterprise companies have a service gateway that requires an auth headers. We can make the desktop browser RemoteWebDriver requests pass the gateway with the required headers because  RemoteWebDriver allows my CommandExecutor. The same is blocked with Appium RemoteWebDriver. 

Please add additional AppiumDriver const which takes user-provided CommandExecutor.

**Comment by TikhomirovSergey** [comment]
Sounds meaningful :+1:

**Comment by Jonahss** [comment]
Good point. We provide our own command-executor to RemoteWebDriver in order to get appium to work, and the chain shouldn't end with us.

@baechul Do you want to work on this and submit a pull request?

**Comment by baechul** [comment]
I am trying to contribute by sending a pull request but I can't find the contribution guide page. Just running "mvn clean install" failed out of the master branch. I am on osx. Is there contribution guide with build/configuration instruction?

**Comment by TikhomirovSergey** [comment]
I think there there should be one or few public constructors that can pass user's CommandExecutor.

Maybe... Does it make sense to use only [HttpCommandExecutor extensions](https://github.com/SeleniumHQ/selenium/blob/ab1e647d0fc8fc39e6b00ae94321ab228b6728f2/java/client/src/org/openqa/selenium/remote/HttpCommandExecutor.java) instead of [CommandExecutor](https://github.com/SeleniumHQ/selenium/blob/ab1e647d0fc8fc39e6b00ae94321ab228b6728f2/java/client/src/org/openqa/selenium/remote/CommandExecutor.java) implementations? This way the number of constructors can be reduced,

**Comment by baechul** [comment]
I have updates to be contributed back. I tested locally and seems working fine. The way I have updated is to allow for users to pass a custom HttpClient.Factory to AndroidDriver/IOSDriver const. The passed HttpClient.Factory is consumed by the existing AppiumCammandExecutor class. 

Example of AndroidDriver additional const:
public AndroidDriver(URL remoteAddress, HttpClient.Factory httpClientFactory, Capabilities desiredCapabilities)

I will send the pull request soon for the community reviews.
